pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/SegmentFinalizer.java

```diff
@@ -0,0 +1,452 @@
+
+    // Once we pick a winner, the winner may get notified in the next call, so add one hold time plus some.
+    public static final long MAX_TIME_TO_NOTIFY_WINNER_MS = MAX_TIME_TO_PICK_WINNER_MS +
+      SegmentFinalProtocol.MAX_HOLD_TIME_MS + (SegmentFinalProtocol.MAX_HOLD_TIME_MS / 10);
+
+    // Once the winner is notified, the are expected to start right away. At this point, it is the segment commit
+    // time that we need to consider.
+    // We may need to add some time here to allow for getting the lock? For now 0
+    // We may need to add some time for the committer come back to us? For now 0.
+    public static final long MAX_TIME_ALLOWED_TO_COMMIT_MS = MAX_TIME_TO_NOTIFY_WINNER_MS + SegmentFinalProtocol.MAX_SEGMENT_COMMIT_TIME_MS;
+
+    public final Logger LOGGER;
+
+    State _state = State.HOLDING;   // Always start off in HOLDING state.
+    final long _startTime;
+    private final LLCSegmentName _segmentName;
+    private final int _numReplicas;
+    private final Map<String, Long> _commitStateMap;
+    private long _winningOffset = -1L;
+    private String _winner;
+    private final PinotLLCRealtimeSegmentManager _segmentManager;
+    private final SegmentFinalizer _segmentFinalizer;
+
+    public SegmentFinalizerFSM(PinotLLCRealtimeSegmentManager segmentManager, SegmentFinalizer segmentFinalizer,
+        LLCSegmentName segmentName, int numReplicas, long winningOffset) {
+      // Constructor used when we get an event after a segment is committed.
+      this(segmentManager, segmentFinalizer, segmentName, numReplicas);
+      _state = State.COMMITTED;
+      _winningOffset = winningOffset;
+      _winner = "UNKNOWN";
+    }
+
+    @Override
+      public String toString() {
+        return "{" + _segmentName.getSegmentName() + "," + _state + "," + _startTime + "," + _winner + "," + _winningOffset + "}";
+      }
+
+    public SegmentFinalizerFSM(PinotLLCRealtimeSegmentManager segmentManager, SegmentFinalizer segmentFinalizer,
+        LLCSegmentName segmentName, int numReplicas) {
+      _segmentName = segmentName;
+      _numReplicas = numReplicas;
+      _segmentManager = segmentManager;
+      _commitStateMap = new HashMap<>(_numReplicas);
+      _segmentFinalizer = segmentFinalizer;
+      _startTime = _segmentFinalizer.getCurrentTimeMs();
+      LOGGER = LoggerFactory.getLogger("SegmentFinalizerFSM_"  + segmentName.getSegmentName());
+    }
+
+    public boolean isDone() {
+      return _state.equals(State.COMMITTED) || _state.equals(State.ABORTED);
+    }
+
+    public SegmentFinalProtocol.Response segmentConsumed(String instanceId, long offset) {
+      final long now = _segmentFinalizer.getCurrentTimeMs();
+      synchronized (this) {
+        LOGGER.info("Processing segmentConsumed({}, {})", instanceId, offset);
+        _commitStateMap.put(instanceId, offset);
+        SegmentFinalProtocol.Response response = null;
+        switch (_state) {
+          case HOLDING:
```

Comment:
This state machine is quite complex, is there a way to simplify it?

I get the impression some states could be removed from it by using state that's been written to Helix/ZK in order to keep only the necessary parts for the segment completion phase to work. For example, if a client reports its offset but we've already committed the segment to the property store, there doesn't need to be a FSM state for this case; this would remove one state in the state machine as well as remove the need to keep the state machine in memory for an extended period of time.

If there's a diagram of the valid state transitions available on the GitHub wiki, it would also be nice to link it here to help with understanding this part of the protocol.